### DISCUSSION TITLE
Vertical Text Pixelated

### DISCUSSION THREAD

**Discussion by matt998759** [ImageMagick/ImageMagick#6284] [Help]
### ImageMagick version

7.0.11-0

### Operating system, version and so on

ImageMagick Lambda Layer for Amazon Linux 2 AMIs

### Description

We are having an issue with text that is rotated 90 degrees where the intrinsic size is much smaller than the rendered size causing pixelation.  Any idea where to fix this?  Screenshot is attached.

![Screenshot 2023-04-26 072323](https://user-images.githubusercontent.com/131872692/234560653-62a9b69e-98cf-4733-84d5-7449ea44c9b1.png)

**Comment by snibgo** [comment]
Please show the IM command you used.

**Reply by matt998759** [reply]
`// Layers
// - https://serverlessrepo.aws.amazon.com/applications/us-east-1/145266761615/image-magick-lambda-layer
// - https://serverlessrepo.aws.amazon.com/applications/us-east-1/145266761615/rsvg-convert-lambda-layer
// - https://github.com/rpidanny/gm-lambda-layer (only required for "gm" binaries)

const AWS = require("aws-sdk");
const fs = require("fs");
const path = require("path");
const { exec } = require("child_process");
const gm = require("gm").subClass({ imageMagick: true });

const s3 = new AWS.S3({});

exports.handler = async function (event, context) {
  // console.log("## ENVIRONMENT VARIABLES: " + serialize(process.env));
  // console.log("## CONTEXT: " + serialize(context));
  console.log("#### EVENT: " + serialize(event));
  // send the response immediately
  context.callbackWaitsForEmptyEventLoop = false;

  cleanTempFolder();

  let response;
  const { bucket, type, src, dest, options, commands, font } = event;

  try {
    switch (type) {
      case "command":
        response = await convertImage(bucket, src, dest, commands);
        break;
      case "text2png":
        response = await convertText2PNG(bucket, font, options, dest);
        break;
      case "monogram2png":
        response = await convertMonogram2PNG(bucket, options, dest);
        break;
      case "svg2png":
        response = await convertSVG2PNG(bucket, src, dest, commands);
        break;
      default:
        throw new Error("not found type");
    }

    return { success: true };
    // return response;
  } catch (err) {
    throw err;
  }
};

var convertImage = async function (bucket, srcKey, destKey, commands) {
  console.log("[convertImage] start:");
  const filePath = "/tmp/" + path.basename(srcKey);

  const convertStart = process.hrtime();
  try {
    const fileFromS3 = await downloadFromS3(bucket, srcKey);
    fs.writeFileSync(filePath, fileFromS3);
    const commandList = commands.map((command) => {
      var str = command.replace(";", "");
      str = str.replace("#input", filePath);
      str = str.replace("#output", filePath);
      return "convert " + str;
    });
    const joinedCommand = commandList.join("; ");

    const convertedFile = await new Promise(function (resolve, reject) {
      exec(joinedCommand, function (err, stdout, stderr) {
        if (err) {
          return reject(err);
        }
        resolve(fs.readFileSync(filePath));
      });
    });
    await uploadToS3(bucket, destKey, convertedFile);
    console.log("[convertImage] success: ", destKey);
  } catch (err) {
    console.log("[convertImage] error: ", err);
    throw err;
  } finally {
    fs.unlinkSync(filePath);
    const convertEnd = process.hrtime(convertStart);
    const convertTimeMs = convertEnd[0] * 1000 + convertEnd[1] / 1000000;
    console.log("[convertImage] finally: ", convertTimeMs);
  }
};

var convertText2PNG = async function (bucket, fontKey, options, destKey) {
  console.log("[convertText2PNG] start:");
  const fontPath = "/tmp/" + path.basename(fontKey);
  const {
    width,
    height,
    color,
    font_size: fontSize,
    text,
    rotation = 0,
  } = options;

  const convertStart = process.hrtime();
  try {
    const fileFromS3 = await downloadFromS3(bucket, fontKey);
    fs.writeFileSync(fontPath, fileFromS3);
    const convertedFile = await new Promise(function (resolve, reject) {
      gm(width, height, "none")
        .fill(color)
        .font(fontPath, fontSize)
        .drawText(0, 0, text, "center")
        .rotate("none", rotation)
        .toBuffer("PNG", function (err, buffer) {
          if (err) {
            return reject(err);
          } else {
            resolve(buffer);
          }
        });
    });
    await uploadToS3(bucket, destKey, convertedFile, "PNG");
    console.log("[convertText2PNG] success: ", destKey);
  } catch (err) {
    console.log("[convertText2PNG] error: ", err);
    throw err;
  } finally {
    fs.unlinkSync(fontPath);
    const convertEnd = process.hrtime(convertStart);
    const convertTimeMs = convertEnd[0] * 1000 + convertEnd[1] / 1000000;
    console.log("[convertText2PNG] finally: ", convertTimeMs);
  }
};

var convertMonogram2PNG = async function (bucket, options, destKey) {
  console.log("[convertMonogram2PNG] start:");
  const {
    type,
    width,
    height,
    color,
    color2,
    font_size: fontSize,
    font_size2: fontSize2,
    text,
    text2,
    rotation = 0,
  } = options;

  const convertStart = process.hrtime();
  try {
    const monogramGenerated = await new Promise(function (resolve, reject) {
      switch (type) {
        case "circle":
          gm(width, height, "none")
            .fill(color)
            .font("fonts/circle_3_black.ttf", fontSize)
            .drawText(-fontSize / 10, (fontSize * 4) / 5, text)
            .rotate("none", rotation)
            .toBuffer("PNG", function (err, buffer) {
              if (err) {
                return reject(err);
              }
              resolve(buffer);
            });
          break;
        case "name_initial":
          gm(width, height, "none")
            .fill(color2)
            .font("fonts/name_initial_lastname.ttf", fontSize2)
            .drawText(0, 0, text2, "center")
            .fill(color)
            .font("fonts/name_initial_firstname.ttf", fontSize)
            .drawText(0, 0, text, "center")
            .rotate("none", rotation)
            .toBuffer("PNG", function (err, buffer) {
              if (err) {
                return reject(err);
              }
              resolve(buffer);
            });
          break;
        case "vine":
          gm(width, height, "none")
            .fill(color)
            .font("fonts/vine.ttf", fontSize)
            .drawText(0, 0, text, "center")
            .rotate("none", rotation)
            .toBuffer("PNG", function (err, buffer) {
              if (err) {
                return reject(err);
              }
              resolve(buffer);
            });
          break;
        default:
          return reject(new Error("unknown mongoram type"));
      }
    });
    await uploadToS3(bucket, destKey, monogramGenerated, "PNG");
    console.log("[convertMonogram2PNG] success: ", destKey);
  } catch (err) {
    console.log("[convertMonogram2PNG] error: ", err);
    throw err;
  } finally {
    const convertEnd = process.hrtime(convertStart);
    const convertTimeMs = convertEnd[0] * 1000 + convertEnd[1] / 1000000;
    console.log("[convertMonogram2PNG] finally: ", convertTimeMs);
  }
};

var convertSVG2PNG = async function (bucket, srcKey, destKey, commands) {
  console.log("[convertSVG2PNG] start:");
  const srcPath = "/tmp/" + path.basename(srcKey);
  const destPath = "/tmp/" + path.basename(destKey);

  const convertStart = process.hrtime();
  try {
    const fileFromS3 = await downloadFromS3(bucket, srcKey);
    fs.writeFileSync(srcPath, fileFromS3);
    const commandList = commands.map((command) => {
      var str = command.replace(";", "");
      str = str.replace("#input", srcPath);
      str = str.replace("#output", destPath);
      return "/opt/bin/" + str;
    });
    const joinedCommand = commandList.join("; ");

    const convertedFile = await new Promise(function (resolve, reject) {
      exec(joinedCommand, function (err, stdout, stderr) {
        if (err) {
          return reject(err);
        }
        resolve(fs.readFileSync(destPath));
      });
    });
    await uploadToS3(bucket, destKey, convertedFile);

    console.log("[convertSVG2PNG] success: ", destKey);
  } catch (err) {
    console.log("[convertSVG2PNG] error: ", err);
    throw err;
  } finally {
    fs.unlinkSync(srcPath);
    fs.unlinkSync(destPath);
    const convertEnd = process.hrtime(convertStart);
    const convertTimeMs = convertEnd[0] * 1000 + convertEnd[1] / 1000000;
    console.log("[convertSVG2PNG] finally: ", convertTimeMs);
  }
};

var downloadFromS3 = async function (bucket, fileKey) {
  console.log("Download Start: ", bucket, fileKey);
  const downloadStart = process.hrtime();
  const params = {
    Bucket: bucket,
    Key: fileKey,
  };

  try {
    const response = await s3.getObject(params).promise();
    console.log("Download Success: ", response.ContentType);
    return response.Body;
  } catch (err) {
    console.error("Download Error: ", err);
    throw err;
  } finally {
    const downloadEnd = process.hrtime(downloadStart);
    const downloadTimeMs = downloadEnd[0] * 1000 + downloadEnd[1] / 1000000;
    console.log("Download Complete: %dms", downloadTimeMs);
  }
};

var uploadToS3 = async function (
  bucket,
  fileKey,
  fileBuffer,
  contentType = ""
) {
  console.error("Upload start: ", bucket, fileKey);
  const uploadStart = process.hrtime();
  const params = {
    Bucket: bucket,
    Key: fileKey,
    Body: fileBuffer,
    ACL: "public-read",
    ContentType: contentType,
  };

  try {
    const response = await s3.putObject(params).promise();
    console.log("Upload success: ", response);
    return true;
  } catch (err) {
    console.error("Upload error: ", err);
    return false;
  } finally {
    const uploadEnd = process.hrtime(uploadStart);
    const uploadTimeMs = uploadEnd[0] * 1000 + uploadEnd[1] / 1000000;
    console.log("Upload finally: %dms", uploadTimeMs);
  }
};

var cleanTempFolder = function () {
  console.log("[cleanTempFolder] start: ");
  const tempFolder = "/tmp";

  try {
    fs.readdirSync(tempFolder).forEach((file) => {
      fs.unlinkSync(path.join(tempFolder, file));
      console.log("[cleanTempFolder] deleted: ", path.join(tempFolder, file));
    });
  } catch (err) {
    console.log("[cleanTempFolder] error: ", err);
  } finally {
    console.log("[cleanTempFolder] finally: ");
  }
};

var serialize = function (object) {
  return JSON.stringify(object, null, 2);
};
`

**Reply by matt998759** [reply]
Is that what you were requesting or something else @snibgo ?  Thanks in advance for helping out.

**Comment by snibgo** [comment]
That doesn't really help. I thought your example might have been created by a simple ImageMagick command.

I am trying to understand your problem. You say "where the intrinsic size is much smaller than the rendered size causing pixelation." I don't know what you mean by "intrinsic size" or "rendered size".

**Reply by matt998759** [reply]
By intrinsic size I mean the size the actual image is where the rendered size is the size it should be.  It only happens when the text is rotated but the actual image size is much smaller than it should be which results in significant pixelation.

**Comment by snibgo** [comment]
Rotating text won't change the size, or pixelate it.

```
magick -size 300x300 xc:Blue -pointsize 25 -draw "text 100,100 Hello" h1.png
```
![h1](https://user-images.githubusercontent.com/33812368/234634314-864a3975-6f7c-440f-8bd3-a0acff7ce945.png)

```
magick -size 300x300 xc:Blue -pointsize 25 -draw "text 100,100 Hello" -rotate -90 h2.png
```

![h2](https://user-images.githubusercontent.com/33812368/234634351-5037bd85-c392-4bc4-b007-b76dc548d426.png)


Enlarging text will change the size, and make it blurry or pixelated or both.

```
magick -size 60x60 xc:Blue -pointsize 5 -draw "text 20,20 Hello" -resize 500% h3.png
```

![h3](https://user-images.githubusercontent.com/33812368/234634400-66e00290-f325-4e20-9a36-f4495923d16b.png)


```
magick -size 60x60 xc:Blue -pointsize 5 -draw "text 20,20 Hello" -scale 500% h4.png
```
![h4](https://user-images.githubusercontent.com/33812368/234634435-0eed6ccc-e129-4b37-9928-038dcaeba3c3.png)
